// 
// This defines the gpb message format used by the npu utilization sensor
//  
// The top level messages is NetworkProcessorUtilization
//  
// Copyrights (c)  2015, Juniper Networks, Inc.
// All rights reserved.
//  
    
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//  
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
//
// Nitin Kumar, Jan 2015
//  
// This file defines the messages in Protocol Buffers format used by
// the Network Processor Utilization sensor. 
//
// Version 1.0
//
syntax = "proto2";
import "jvision_top.proto";
    
//  
// This occupies branch 11 from JuniperNetworksSensors
//
extend JuniperNetworksSensors {
    optional NetworkProcessorUtilization jnpr_npu_util_ext        = 11;
}

// Top level message
message NetworkProcessorUtilization {
    repeated Utilization npu_util_stats = 1;
}

// Utilization per Network Processor
message Utilization {
    // Globally unique identifier for an NPU. This is of the form
    // "pfe-X/Y", where X is the slot number of the linecard and Y
    // is the index of the NPU on the linecard
    required string        identifier        = 1; 

    // A number on a scale of 0-100 that indicates the business of 
    // an NPU. This is an approximation as the utilization depends
    // on the busyness of several internal components of the NPU
    optional uint32        utilization       = 2;  

    // Offered Packet load on the NPU. 
    repeated PacketLoad    packets           = 3;

    // Utilization of various internal memories of the NPU. 
    repeated MemoryLoad    memory            = 4;
};

// Load on a memory subsystem of the NPU
message MemoryLoad {
    // A name string to identify the particular memory subsystem
    optional string name                     = 1;

    // Various utilization metrics over a period of last 3 seconds
    optional uint32 average_util             = 2;
    optional uint32 highest_util             = 3;
    optional uint32 lowest_util              = 4;

    // Each memory is front ended by a cache. The following metrics 
    // indicate how these caches are working
    optional uint32 average_cache_hit_rate   = 5;
    optional uint32 highest_cache_hit_rate   = 6;
    optional uint32 lowest_cache_hit_rate    = 7;
}

// Offered packet load on an internal subsytem of the NPU
message PacketLoad {
    // Each internal subsystem of the NPU has a name
    required string identifier = 1;

    // Rate of packets received
    optional uint32 rate       = 2;

    // The following metrics indicate the compute load on the NPU
    optional uint32 average_instructions_per_packet = 3;
    optional uint32 average_wait_cycles_per_packet  = 4;
    optional uint32 average_cycles_per_packet       = 5;
}
